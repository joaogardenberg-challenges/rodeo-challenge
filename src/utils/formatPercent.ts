export default function formatPercent(price: number) {
  return price.toLocaleString('en', {
    minimumFractionDigits: 0,
    maximumFractionDigits: 2
  })
}
