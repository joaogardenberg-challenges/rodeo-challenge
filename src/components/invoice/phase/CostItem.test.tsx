import { useMemo } from 'react'
import { render, screen } from '@testing-library/react'
import { noop } from 'lodash'
import formatPrice from 'utils/formatPrice'
import formatPercent from 'utils/formatPercent'

const useContext = jest.fn()
const getCostItemPrice = jest.fn()
jest.doMock('react', () => ({ useContext, useMemo }))
jest.doMock('utils/costItemPrice', () => ({ getCostItemPrice }))
jest.doMock('components/invoice/Invoice', () => ({ InvoiceContext: noop }))

const CostItem = require('./CostItem').default

const name = 'Cost item name'
const tax = 9
const price = 10
const totalPrice = 100
const costItem = { name, tax, price }

describe('CostItem', () => {
  beforeEach(() => {
    useContext.mockReturnValue({ currency: 'USD' })
    getCostItemPrice.mockReturnValue(totalPrice)
  })

  it('Renders successfully with 1 unit', () => {
    const { container } = render(
      <CostItem costItem={{ ...costItem, billedPerUnit: true, units: 1 }} />
    )

    const nameElement = screen.getByTestId('name')
    const taxElement = screen.getByTestId('tax')
    const description = screen.getByTestId('description')
    const priceElement = screen.getByTestId('price')

    expect(nameElement).toHaveTextContent(name)
    expect(taxElement).toHaveTextContent(`${formatPercent(tax)}%`)
    expect(description).toHaveTextContent(`$${formatPrice(price)} x 01 unit`)
    expect(priceElement).toHaveTextContent(`$${formatPrice(totalPrice)}`)
    expect(container).toMatchSnapshot()
  })

  it('Renders successfully with 1 hour', () => {
    const { container } = render(
      <CostItem costItem={{ ...costItem, billedPerUnit: false, units: 1 }} />
    )

    const nameElement = screen.getByTestId('name')
    const taxElement = screen.getByTestId('tax')
    const description = screen.getByTestId('description')
    const priceElement = screen.getByTestId('price')

    expect(nameElement).toHaveTextContent(name)
    expect(taxElement).toHaveTextContent(`${formatPercent(tax)}%`)
    expect(description).toHaveTextContent(`$${formatPrice(price)} x 01 hour`)
    expect(priceElement).toHaveTextContent(`$${formatPrice(totalPrice)}`)
    expect(container).toMatchSnapshot()
  })

  it('Renders successfully with multiple units', () => {
    const units = 10

    const { container } = render(
      <CostItem costItem={{ ...costItem, billedPerUnit: true, units }} />
    )

    const nameElement = screen.getByTestId('name')
    const taxElement = screen.getByTestId('tax')
    const description = screen.getByTestId('description')
    const priceElement = screen.getByTestId('price')

    expect(nameElement).toHaveTextContent(name)
    expect(taxElement).toHaveTextContent(`${formatPercent(tax)}%`)

    expect(description).toHaveTextContent(
      `$${formatPrice(price)} x ${units} units`
    )

    expect(priceElement).toHaveTextContent(`$${formatPrice(totalPrice)}`)
    expect(container).toMatchSnapshot()
  })

  it('Renders successfully with multiple hours', () => {
    const hours = 10

    const { container } = render(
      <CostItem costItem={{ ...costItem, billedPerUnit: false, hours }} />
    )

    const nameElement = screen.getByTestId('name')
    const taxElement = screen.getByTestId('tax')
    const description = screen.getByTestId('description')
    const priceElement = screen.getByTestId('price')

    expect(nameElement).toHaveTextContent(name)
    expect(taxElement).toHaveTextContent(`${formatPercent(tax)}%`)

    expect(description).toHaveTextContent(
      `$${formatPrice(price)} x ${hours} hours`
    )

    expect(priceElement).toHaveTextContent(`$${formatPrice(totalPrice)}`)
    expect(container).toMatchSnapshot()
  })
})
