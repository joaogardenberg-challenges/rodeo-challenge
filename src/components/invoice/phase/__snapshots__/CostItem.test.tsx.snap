// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CostItem Renders successfully with 1 hour 1`] = `
<div>
  <li
    class="cost-item"
  >
    <span
      hidden=""
    >
      Name: 
    </span>
    <span
      class="cost-item__name"
      data-testid="name"
    >
      Cost item name
    </span>
    <span
      hidden=""
    >
      , Tax: 
    </span>
    <span
      class="cost-item__tax"
      data-testid="tax"
    >
      9
      %
    </span>
    <span
      hidden=""
    >
      , Description: 
    </span>
    <span
      class="cost-item__description"
      data-testid="description"
    >
      $
      10.00
       x 
      01
       
      hour
    </span>
    <span
      hidden=""
    >
      , Price: 
    </span>
    <span
      class="cost-item__price"
      data-testid="price"
    >
      $
      100.00
    </span>
  </li>
</div>
`;

exports[`CostItem Renders successfully with 1 unit 1`] = `
<div>
  <li
    class="cost-item"
  >
    <span
      hidden=""
    >
      Name: 
    </span>
    <span
      class="cost-item__name"
      data-testid="name"
    >
      Cost item name
    </span>
    <span
      hidden=""
    >
      , Tax: 
    </span>
    <span
      class="cost-item__tax"
      data-testid="tax"
    >
      9
      %
    </span>
    <span
      hidden=""
    >
      , Description: 
    </span>
    <span
      class="cost-item__description"
      data-testid="description"
    >
      $
      10.00
       x 
      01
       
      unit
    </span>
    <span
      hidden=""
    >
      , Price: 
    </span>
    <span
      class="cost-item__price"
      data-testid="price"
    >
      $
      100.00
    </span>
  </li>
</div>
`;

exports[`CostItem Renders successfully with multiple hours 1`] = `
<div>
  <li
    class="cost-item"
  >
    <span
      hidden=""
    >
      Name: 
    </span>
    <span
      class="cost-item__name"
      data-testid="name"
    >
      Cost item name
    </span>
    <span
      hidden=""
    >
      , Tax: 
    </span>
    <span
      class="cost-item__tax"
      data-testid="tax"
    >
      9
      %
    </span>
    <span
      hidden=""
    >
      , Description: 
    </span>
    <span
      class="cost-item__description"
      data-testid="description"
    >
      $
      10.00
       x 
      10
       
      hours
    </span>
    <span
      hidden=""
    >
      , Price: 
    </span>
    <span
      class="cost-item__price"
      data-testid="price"
    >
      $
      100.00
    </span>
  </li>
</div>
`;

exports[`CostItem Renders successfully with multiple units 1`] = `
<div>
  <li
    class="cost-item"
  >
    <span
      hidden=""
    >
      Name: 
    </span>
    <span
      class="cost-item__name"
      data-testid="name"
    >
      Cost item name
    </span>
    <span
      hidden=""
    >
      , Tax: 
    </span>
    <span
      class="cost-item__tax"
      data-testid="tax"
    >
      9
      %
    </span>
    <span
      hidden=""
    >
      , Description: 
    </span>
    <span
      class="cost-item__description"
      data-testid="description"
    >
      $
      10.00
       x 
      10
       
      units
    </span>
    <span
      hidden=""
    >
      , Price: 
    </span>
    <span
      class="cost-item__price"
      data-testid="price"
    >
      $
      100.00
    </span>
  </li>
</div>
`;
